cmd/startup/startup.go

```diff
@@ -0,0 +1,198 @@
+	"github.com/openshift/openshift-azure/pkg/util/template"
+)
+
+var (
+	logLevel  = flag.String("loglevel", "Debug", "valid values are Debug, Info, Warning, Error")
+	gitCommit = "unknown"
+)
+
+type startup struct {
+	blob azureclientstorage.Blob
+	log  *logrus.Entry
+}
+
+func (s *startup) writeTemplatedFiles(cs *api.OpenShiftManagedCluster) error {
+	tc := api.TestConfig{RunningUnderTest: cs.Config.RunningUnderTest}
+	hostname, _ := os.Hostname()
+	cname, _ := net.LookupCNAME(hostname)
+	domainname := strings.SplitN(strings.TrimSuffix(cname, "."), ".", 2)[1]
+
+	s.log.Infof("hostname is \"%s\"", hostname)
+	s.log.Infof("domainname %s", domainname)
+
+	files := []string{
+		"etc/etcd/ca.crt",
+		"etc/etcd/server.crt",
+		"etc/etcd/server.key",
+		"etc/etcd/peer.crt",
+		"etc/etcd/peer.key",
+		"etc/origin/node/ca.crt",
+		"etc/origin/master/master.etcd-ca.crt",
+		"etc/origin/master/ca.crt",
+		"etc/origin/master/ca.key",
+		"etc/origin/master/named/console.crt",
+		"etc/origin/master/named/console.key",
+		"etc/origin/master/front-proxy-ca.crt",
+		"etc/origin/master/service-signer.crt",
+		"etc/origin/master/service-signer.key",
+		"etc/origin/master/master.etcd-client.crt",
+		"etc/origin/master/master.etcd-client.key",
+		"etc/origin/master/aggregator-front-proxy.crt",
+		"etc/origin/master/aggregator-front-proxy.key",
+		"etc/origin/master/master.kubelet-client.crt",
+		"etc/origin/master/master.kubelet-client.key",
+		"etc/origin/master/master.proxy-client.crt",
+		"etc/origin/master/master.proxy-client.key",
+		"etc/origin/master/master.server.crt",
+		"etc/origin/master/master.server.key",
+		"etc/origin/master/serviceaccounts.public.key",
+		"etc/origin/master/serviceaccounts.private.key",
+		"etc/etcd/etcd.conf",
+		"etc/origin/master/master-config.yaml",
+		"etc/origin/master/session-secrets.yaml",
+		"etc/origin/master/scheduler.json",
+		"etc/origin/node/pods/etcd.yaml",
+		"etc/origin/node/pods/apiserver.yaml",
+		"etc/origin/node/pods/controllers.yaml",
+		"etc/origin/node/node.kubeconfig",
+		"etc/origin/master/admin.kubeconfig",
+		"etc/origin/master/openshift-master.kubeconfig",
+	}
+	if hostname == "master-000000" {
+		files = append(files, "etc/origin/node/pods/sync.yaml")
+	}
+
+	for _, templateFileName := range files {
+		s.log.Debugf("processing template %s", templateFileName)
+		templateFile, err := arm.Asset(templateFileName)
+		if err != nil {
+			return errors.Wrapf(err, "Asset(%s)", templateFileName)
+		}
+		b, err := template.Template(string(templateFile), nil, cs, map[string]interface{}{
+			"Hostname":    hostname,
+			"DNSHostname": domainname,
+			"TestConfig":  tc,
+		})
+		if err != nil {
+			return errors.Wrapf(err, "Template(%s)", templateFileName)
+		}
+		parentDir := path.Base("/" + templateFileName)
+		err = os.MkdirAll(parentDir, 0700)
+		if err != nil {
+			return errors.Wrapf(err, "MkdirAll(%s)", parentDir)
+		}
+		err = ioutil.WriteFile("/"+templateFileName, b, 0600)
```

Comment:
0600 if it's a key or kubeconfig; 0666 otherwise (your uname will take that down to 0644 or whatever)

Reply:
TestConfig is used

```
{{- if $.Extra.TestConfig.RunningUnderTest }}
    - --loglevel=4
{{- else }}
    - --loglevel=2
{{- end }}

```

Reply:
If you rewrite `.Extra.TestConfig.RunningUnderTest` to `.Config.RunningUnderTest`, you don't need `TestConfig` any more.